File: language/runtime/apple/core/Sources/RuntimeHostAppleCore/Module/Calendar/CalendarRequests.swift
import Foundation

/// The Apple calendar request surface for one runtime host embedder.
@MainActor
public protocol CalendarRequests: AnyObject {
    /// List the readable host calendars.
    func listCalendars() -> (status: UInt32, calendars: [RuntimeHostCalendarDescriptor]?)

    /// List host calendar events for one query.
    func listCalendarEvents(
        _ query: RuntimeHostCalendarEventQuery
    ) -> (status: UInt32, events: [RuntimeHostCalendarEvent]?)

    /// Read one host calendar event by stable identifier.
    func readCalendarEvent(
        id: String
    ) -> (status: UInt32, event: RuntimeHostCalendarEvent?)

    /// Create one host calendar event and return its identifier.
    func createCalendarEvent(
        _ draft: RuntimeHostCalendarEventDraft
    ) -> (status: UInt32, id: String?)

    /// Update one existing host calendar event.
    func updateCalendarEvent(
        id: String,
        draft: RuntimeHostCalendarEventDraft
    ) -> UInt32

    /// Delete one existing host calendar event.
    func deleteCalendarEvent(
        id: String
    ) -> UInt32
}

/// The explicit unsupported calendar request surface for one Apple runtime host.
@MainActor
public final class UnsupportedCalendarRequests: CalendarRequests {
    /// Create one unsupported calendar request surface.
    public init() {}

    public func listCalendars() -> (status: UInt32, calendars: [RuntimeHostCalendarDescriptor]?) {
        (hostStatusNotSupported, nil)
    }

    public func listCalendarEvents(
        _ query: RuntimeHostCalendarEventQuery
    ) -> (status: UInt32, events: [RuntimeHostCalendarEvent]?) {
        let _ = query

        return (hostStatusNotSupported, nil)
    }

    public func readCalendarEvent(
        id: String
    ) -> (status: UInt32, event: RuntimeHostCalendarEvent?) {
        let _ = id

        return (hostStatusNotSupported, nil)
    }

    public func createCalendarEvent(
        _ draft: RuntimeHostCalendarEventDraft
    ) -> (status: UInt32, id: String?) {
        let _ = draft

        return (hostStatusNotSupported, nil)
    }

    public func updateCalendarEvent(
        id: String,
        draft: RuntimeHostCalendarEventDraft
    ) -> UInt32 {
        let _ = id
        let _ = draft

        return hostStatusNotSupported
    }

    public func deleteCalendarEvent(
        id: String
    ) -> UInt32 {
        let _ = id

        return hostStatusNotSupported
    }
}

File: language/runtime/android/kotlin/src/main/kotlin/dev/destack/runtime/android/module/calendar/CalendarRequests.kt
package dev.destack.runtime.android.module.calendar

import dev.destack.runtime.android.core.hostStatusNotSupported

/**
 * The calendar request surface for one Android runtime host.
 */
public interface CalendarRequests {
    /**
     * List readable calendars through the host.
     */
    public fun listCalendars(): RuntimeHostCalendarListResponse

    /**
     * List calendar events through the host.
     */
    public fun listCalendarEvents(
        query: RuntimeHostCalendarEventQuery,
    ): RuntimeHostCalendarEventListResponse

    /**
     * Read one calendar event by stable identifier.
     */
    public fun readCalendarEvent(
        id: String,
    ): RuntimeHostCalendarEventResponse

    /**
     * Create one calendar event and return its stable identifier.
     */
    public fun createCalendarEvent(
        draft: RuntimeHostCalendarEventDraft,
    ): RuntimeHostCalendarEventCreateResponse

    /**
     * Update one calendar event by stable identifier.
     */
    public fun updateCalendarEvent(
        id: String,
        draft: RuntimeHostCalendarEventDraft,
    ): Int

    /**
     * Delete one calendar event by stable identifier.
     */
    public fun deleteCalendarEvent(
        id: String,
    ): Int
}

/**
 * The explicit unsupported calendar request surface for one Android runtime host.
 */
public object UnsupportedCalendarRequests : CalendarRequests {
    override fun listCalendars(): RuntimeHostCalendarListResponse {
        return RuntimeHostCalendarListResponse(status = hostStatusNotSupported)
    }

    override fun listCalendarEvents(
        query: RuntimeHostCalendarEventQuery,
    ): RuntimeHostCalendarEventListResponse {
        return RuntimeHostCalendarEventListResponse(status = hostStatusNotSupported)
    }

    override fun readCalendarEvent(
        id: String,
    ): RuntimeHostCalendarEventResponse {
        return RuntimeHostCalendarEventResponse(status = hostStatusNotSupported)
    }

    override fun createCalendarEvent(
        draft: RuntimeHostCalendarEventDraft,
    ): RuntimeHostCalendarEventCreateResponse {
        return RuntimeHostCalendarEventCreateResponse(status = hostStatusNotSupported)
    }

    override fun updateCalendarEvent(
        id: String,
        draft: RuntimeHostCalendarEventDraft,
    ): Int {
        return hostStatusNotSupported
    }

    override fun deleteCalendarEvent(
        id: String,
    ): Int {
        return hostStatusNotSupported
    }
}
